### DISCUSSION TITLE
OData batching does not deserialize complex types in controller post actions

### DISCUSSION THREAD

**Discussion by levelnis** [OData/AspNetCoreOData#1172] [Q&A] [answered]
I'm using the `DefaultODataBatchHandler` to process batch requests, but am having problems with the resultant controller action. The requests go to 3 different controller actions. Each accepts a complex type as a parameter. I've inspected the request stream to ensure that the correct JSON exists to hydrate the target types, but I'm either getting a null value if the action parameter uses the `FromBody` attribute, or an object with default values, but no values populated from the request. 

Here's an example of the request json:

```
{
  "requests": [
    {
      "id": "1",
      "method": "POST",
      "url": "/api/things/1a87f869-5f07-4d3d-be42-bf1500379b68/items",
      "headers":
      {
        "content-type": "application/json; odata.metadata=minimal; odata.streaming=true",
        "odata-version": "4.0"
      },
      "body":
      {
        "id": "1",
        "description": "thing"
      }
    }
  ]
}
```

Here's the post action that gets executed by the request delegate. The `MyModels.Item` object is null if `FromBody` is included. If it's not included, `item` exists but the `Description` and `Id` properties are both empty:

```
[HttpPost("{id:guid}/items")]
[Consumes("application/json")]
public async Task<IActionResult> PostItemAsync([FromRoute] Guid id, [FromBody] MyModels.Item item, ODataQueryOptions<MyModels.Item> options)
```

**Comment by gathogojr** [comment, marked as answer]
@levelnis I was unable to reproduce the issue with the following minimal repro:
```csharp
// Data models
namespace NS.Models
{
    public class Thing
    {
        public int Id { get; set; }
        public List<Item> Items { get; set; }
    }

    public class Item
    {
        public int Id { get; set; }
        public string Description { get; set; }
    }
}

// Controller
public class ThingsController : ODataController
{
    [HttpPost("Things/{id:guid}/Items")]
    [Consumes("application/json")]
    public async Task<IActionResult> PostItemAsync([FromRoute] Guid id, [FromBody] NS.Models.Item item, ODataQueryOptions<NS.Models.Item> options)
    {
        return Ok();
    }
}

// Service configuration
using Microsoft.AspNetCore.OData;
using Microsoft.AspNetCore.OData.Batch;
using Microsoft.OData.ModelBuilder;

var builder = WebApplication.CreateBuilder(args);

var modelBuilder = new ODataConventionModelBuilder();
modelBuilder.EntitySet<Thing>("Things");
modelBuilder.ComplexType<Item>();

builder.Services.AddControllers().AddOData(
    options => options.EnableQueryFeatures().AddRouteComponents(
        routePrefix: string.Empty,
        model: modelBuilder.GetEdmModel(),
        batchHandler: new DefaultODataBatchHandler()));

var app = builder.Build();

app.UseODataBatching();
app.UseRouting();
app.MapControllers();

app.Run();
```

If I hit a service based on the above with the following payload mirroring yours, everything works okay:
```json
{
  "requests": [
    {
      "id": "1",
      "method": "POST",
      "url": "/Things/1a87f869-5f07-4d3d-be42-bf1500379b68/Items",
      "headers":
      {
        "Content-Type": "application/json; odata.metadata=minimal; odata.streaming=true",
        "OData-Version": "4.0"
      },
      "body":
      {
        "Id": "1",
        "Description": "thing"
      }
    }
  ]
}
```

Maybe you can share a repro to help us understand the issue.

**Reply by levelnis** [reply]
Hey @gathogojr , thanks for taking the time to look at this. I took your advice and tried to create a simple example so I could confirm the functionality. I did and it worked as expected. I then went back to the endpoints that had the original problem and they worked as expected as well. So it's actually fine! 

Not sure what happened there or how I fixed it. We use a lot of OData customisations, so creating a simple example to highlight this behaviour was a little tricky with our set up. Chances are I did something that fixed the issue, not sure what though. I'll mark your response as an answer because we can close this.

**Reply by gathogojr** [reply]
Great! I'm glad to have been of some help
